From: IrfanMohammad <irfan@graphiant.com>
Date: Sun, 29 Oct 2023 14:10:33 +0000
Subject: bugfix edit_diff add followed by del

When delete follows a set in an edit, they are both deleted. However this
is not correct. The delete should act on all previous sets, and also on
the stored data.

Closes: ENG-5756
---
 src/edit_diff.c        | 17 +++++++++++------
 src/sysrepo.c          | 10 ++++++----
 tests/test_oper_push.c | 38 ++++++++++++++++++++++++++++++++++----
 3 files changed, 51 insertions(+), 14 deletions(-)

diff --git a/src/edit_diff.c b/src/edit_diff.c
index bc4810b..1e0edc6 100644
--- a/src/edit_diff.c
+++ b/src/edit_diff.c
@@ -3079,7 +3079,7 @@ sr_edit_add_descendant_have_own_op(const struct lyd_node *sibling, enum edit_op
  * @return err_info, NULL on success.
  */
 static sr_error_info_t *
-sr_edit_add_merge_op(struct lyd_node *match, struct lyd_node **root, const char *value, enum edit_op op, int oper_ds)
+sr_edit_add_merge_op(struct lyd_node *match, struct lyd_node **root, const char *value, enum edit_op op, int oper_ds, int *same_op)
 {
     sr_error_info_t *err_info = NULL;
     const struct lysc_node *schema;
@@ -3087,10 +3087,11 @@ sr_edit_add_merge_op(struct lyd_node *match, struct lyd_node **root, const char
     union lyd_any_value any_val;
     enum edit_op cur_op;
     int own_oper;
-
+    *same_op = 0;
     cur_op = sr_edit_diff_find_oper(match, 1, &own_oper);
     if (op == cur_op) {
         /* same node with same operation, silently ignore */
+        *same_op = 1;
         goto cleanup;
     }
 
@@ -3613,12 +3614,13 @@ sr_edit_add(sr_session_ctx_t *session, const char *xpath, const char *value, con
     const char *meta_val = NULL, *def_origin;
     char *rel_xpath = NULL;
     enum edit_op op;
-    int opts, own_oper;
-
+    int opts, own_oper, same_op = 0;
+    int removal = 0;
     assert(!origin || strchr(origin, ':'));
 
     opts = LYD_NEW_PATH_WITH_OPAQ;
     if (!strcmp(operation, "remove") || !strcmp(operation, "delete") || !strcmp(operation, "purge")) {
+        removal = 1;
         opts |= LYD_NEW_PATH_OPAQ;
     }
 
@@ -3631,10 +3633,13 @@ sr_edit_add(sr_session_ctx_t *session, const char *xpath, const char *value, con
         if (match) {
             /* node exists, nothing to create, just merge operations if possible */
             if ((err_info = sr_edit_add_merge_op(match, &session->dt[session->ds].edit->tree, value,
-                    sr_edit_str2op(operation), (session->ds == SR_DS_OPERATIONAL)))) {
+                    sr_edit_str2op(operation), (session->ds == SR_DS_OPERATIONAL), &same_op))) {
                 goto error_safe;
             }
-            goto success;
+            /* We want to delete not just what's in the edit but what's in datastore as well */
+            if (same_op || !removal) {
+                goto success;
+            }
         }
     }
 
diff --git a/src/sysrepo.c b/src/sysrepo.c
index c3d1c0e..d60d446 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -3437,16 +3437,18 @@ cleanup:
 }
 
 API int
-sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_options_t opts)
+sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_options_t in_opts)
 {
     sr_error_info_t *err_info = NULL;
     const char *operation;
     const struct lysc_node *snode;
     uint32_t temp_lo = 0;
-    int isolate = SR_EDIT_ISOLATE & opts;
 
-    if (isolate && SR_DS_OPERATIONAL == session->ds) {
-        return sr_discard_items(session, path);
+    sr_edit_options_t opts = in_opts;
+
+    if (session && session->ds == SR_DS_OPERATIONAL) {
+        /* operational datastore does not support isolate flag */
+        opts &= ~SR_EDIT_ISOLATE;
     }
 
     SR_CHECK_ARG_APIRET(!session || !path || SR_EDIT_DS_API_CHECK(session->ds, opts), session, err_info);
diff --git a/tests/test_oper_push.c b/tests/test_oper_push.c
index effc26b..1dd0899 100644
--- a/tests/test_oper_push.c
+++ b/tests/test_oper_push.c
@@ -159,11 +159,14 @@ clear_up(void **state)
     return 0;
 }
 
+/* TEST */
 static void
 test_oper_set_del_new(void **state)
 {
     struct state *st = (struct state *)*state;
     int i, ret;
+    sr_data_t *data = NULL;
+    char *str1 = NULL, *str2 = NULL;
 
     char *xpath_set[] = {
         "/ietf-interfaces:interfaces-state/interface[name='eth1']/type",
@@ -183,18 +186,45 @@ test_oper_set_del_new(void **state)
         NULL
     };
 
+    /* switch to operational DS */
+    ret = sr_session_switch_ds(st->sess, SR_DS_OPERATIONAL);
+    assert_int_equal(ret, SR_ERR_OK);
+
     for (i = 0; xpath_set[i]; i++) {
-        ret = sr_set_item_str(st->sess, xpath_set[i], xpath_set_vals[i], NULL, SR_EDIT_ISOLATE);
+        ret = sr_set_item_str(st->sess, xpath_set[i], xpath_set_vals[i], NULL, 0);
         assert_int_equal(ret, SR_ERR_OK);
     }
-    assert_int_equal(ret, SR_ERR_OK);
-
     for (i = 0; xpath_del[i]; i++) {
-        ret = sr_delete_item(st->sess, xpath_del[i], SR_EDIT_ISOLATE);
+        ret = sr_delete_item(st->sess, xpath_del[i], /* SR_EDIT_ISOLATE */ 0);
+        assert_int_equal(ret, SR_ERR_OK);
+    }
+
+    for (i = 0; xpath_set[i]; i++) {
+        ret = sr_set_item_str(st->sess, xpath_set[i], xpath_set_vals[i], NULL, 0);
         assert_int_equal(ret, SR_ERR_OK);
     }
+
     ret = sr_apply_changes(st->sess, 0);
     assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_get_data(st->sess, "/ietf-interfaces:interfaces-state", 0, 0, 0, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ret = lyd_print_mem(&str1, data->tree, LYD_XML, LYD_PRINT_WITHSIBLINGS);
+    assert_int_equal(ret, 0);
+
+    str2 = "<interfaces-state xmlns=\"urn:ietf:params:xml:ns:yang:ietf-interfaces\">\n"
+        "  <interface>\n"
+        "    <name>eth1</name>\n"
+        "    <type xmlns:ianaift=\"urn:ietf:params:xml:ns:yang:iana-if-type\">ianaift:ethernetCsmacd</type>\n"
+        "    <oper-status>down</oper-status>\n"
+        "  </interface>\n"
+        "</interfaces-state>\n";
+
+    assert_string_equal(str1, str2);
+    free(str1);
+
+    sr_release_data(data);
 }
 
 static int
